iree/runtime/session.h

```diff
@@ -0,0 +1,224 @@
+
+// Builtin modules that are provided by the runtime.
+enum iree_runtime_session_builtins_e {
+  // All built-in modules that are compiled into the runtime will be available.
+  IREE_RUNTIME_SESSION_BUILTIN_ALL = UINT64_MAX,
+};
+typedef uint64_t iree_runtime_session_builtins_t;
+
+// Options used to configure session creation.
+typedef struct {
+  // A bitmask identifying which IREE builtin modules should be enabled.
+  // Session creation will fail if a requested module is not built into the
+  // runtime binary.
+  iree_runtime_session_builtins_t builtin_modules;
+} iree_runtime_session_options_t;
+
+// Initializes |out_options| to its default values.
+IREE_API_EXPORT void iree_runtime_session_options_initialize(
+    iree_runtime_session_options_t* out_options);
+
+//===----------------------------------------------------------------------===//
+// iree_runtime_session_t
+//===----------------------------------------------------------------------===//
+
+// Creates a new session forced to use the given |device|.
+// This bypasses any device enumeration performed by the loaded modules but
+// the loaded modules will still verify that the device matches their
+// requirements.
+//
+// A base set of modules may be added by the runtime during creation based on
+// |options| and users may load additional modules - such as the one containing
+// their user code - by using the iree_vm_context_t provided by
+// iree_runtime_session_context.
+//
+// |host_allocator| will be used to allocate the session and any associated
+// resources. |out_session| must be released by the caller.
+IREE_API_EXPORT iree_status_t iree_runtime_session_create_with_device(
+    iree_runtime_instance_t* instance,
+    const iree_runtime_session_options_t* options, iree_hal_device_t* device,
+    iree_allocator_t host_allocator, iree_runtime_session_t** out_session);
+
+// Retains the given |session| for the caller.
+IREE_API_EXPORT void iree_runtime_session_retain(
+    iree_runtime_session_t* session);
+
+// Releases the given |session| from the caller.
+IREE_API_EXPORT void iree_runtime_session_release(
+    iree_runtime_session_t* session);
+
+// Returns the host allocator used to allocate the session and its resources.
+// Callers should use this to allocate resources so that any memory tracking
+// being performed correctly attributes the allocations to the session.
+IREE_API_EXPORT iree_allocator_t
+iree_runtime_session_host_allocator(const iree_runtime_session_t* session);
+
+// Returns the instance the session uses for shared resources.
+IREE_API_EXPORT iree_runtime_instance_t* iree_runtime_session_instance(
+    const iree_runtime_session_t* session);
+
+// Returns the VM context used to load and link modules.
+// The context can be used to perform additional reflection over the loaded
+// modules or load additional modules (if supported).
+IREE_API_EXPORT iree_vm_context_t* iree_runtime_session_context(
+    const iree_runtime_session_t* session);
```

Comment:
Ok, I was wondering about this.

The higher level API is built on top of the low level API, but still offers access to the underlying structures. So it's possible to use the high level API to get set up then use the low level structures for full expressivity?

Reply:
Yep! So registering custom modules is just getting the context and calling register_module on it without the need for additional plumbing here:
```c
iree_vm_context_register_modules(iree_runtime_session_context(session), ...);
```

Avoids a lot of duplication and potential for errors to be introduced in the plumbing.